lib/streamlit/web/server/routes.py

```diff
@@ -185,6 +185,8 @@ async def get(self):
     "https://*.head.streamlit.app",
     "https://*.staging.streamlit.app",
     "https://*.streamlit.app",
+    # Allow messages from localhost in dev mode for testing of host <-> guest communication
+    "http://localhost" if config.get_option("global.developmentMode") else None,
```

Comment:
There's a very subtle issue with writing this in this way (that we should really figure out a better way of catching aside from hoping to notice it in code review 🙈): the call to `config.get_option` happens when this file is imported, which causes the `config.toml` to be parsed earlier than when we do so during initialization in `load_config_options` in `bootstrap.py`. We need to do this there so that we can also take into account any config options that are set via CLI flag.

I don't think this has any negative effect for the user, but it does mean we parse the `config.toml` file twice during initialization, which is extra work.

This is probably most easily solvable by adding an `initialize` method to the `AllowedMessageOriginsHandler` and making a copy of `ALLOWED_MESSAGE_ORIGINS` there, conditionally adding `localhost` depending on the config option value

Then we can serve the copy from the endpoint handler instead of the original list.

Reply:
Thank you! This addressed the weird test failure behavior I was seeing that had me stumped 😅 
Let me know what you think of the revised `routes.py`